File: java17/src/main/java/folio/codinginterview/infrastructure/repository/MarketPriceRepositoryImpl.java
package folio.codinginterview.infrastructure.repository;

import folio.codinginterview.application.repository.MarketPriceRepository;
import folio.codinginterview.domain.AppConstants;
import folio.codinginterview.domain.StockSymbol;

import java.math.BigDecimal;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicReference;

public final class MarketPriceRepositoryImpl implements MarketPriceRepository {
    private final AtomicReference<Map<StockSymbol, BigDecimal>> ref =
            new AtomicReference<>(AppConstants.INITIAL_PRICES);

    @Override
    public CompletableFuture<Map<StockSymbol, BigDecimal>> all() {
        return CompletableFuture.completedFuture(ref.get());
    }

    @Override
    public CompletableFuture<Void> update(Map<StockSymbol, BigDecimal> prices) {
        ref.set(Map.copyOf(prices));
        return CompletableFuture.completedFuture(null);
    }
}

File: scala/src/main/scala/folio/codinginterview/infrastructure/repository/MarketPriceRepositoryImpl.scala
package folio.codinginterview.infrastructure.repository

import folio.codinginterview.application.repository.MarketPriceRepository
import folio.codinginterview.domain.AppConstants
import folio.codinginterview.domain.StockSymbol
import java.util.concurrent.atomic.AtomicReference
import scala.concurrent.Future

final class MarketPriceRepositoryImpl extends MarketPriceRepository {
  private val ref: AtomicReference[Map[StockSymbol, BigDecimal]] =
    new AtomicReference(AppConstants.initialPrices)

  override def all(): Future[Map[StockSymbol, BigDecimal]] = Future.successful(ref.get())

  override def update(prices: Map[StockSymbol, BigDecimal]): Future[Unit] = {
    ref.set(prices)
    Future.unit
  }
}
